client/src/Editor/CreateObject.jsx

```diff
@@ -0,0 +1,170 @@
+        throw new Error("allowChangeParentPid and allowNoParentPid cannot both be false when parentPid is empty.");
+    }
+    if (allowNoParentPid && !allowChangeParentPid) {
+        throw new Error("allowNoParentPid=true requires allowChangeParentPid to be true");
+    }
+    const ajax = AjaxHelper.getInstance();
+    const [models, setModels] = useState({});
+    const [selectedModel, setSelectedModel] = useState("");
+    const [results, setResults] = useState("");
+    const [title, setTitle] = useState("");
+    const [parent, setParent] = useState(parentPid);
+    const [noParent, setNoParent] = useState(allowNoParentPid && parentPid === "");
+    const [state, setState] = useState("Inactive");
+
+    const states = ["Active", "Inactive", "Deleted"];
+
+    useEffect(() => {
+        ajax.getJSONPromise(ajax.apiUrl + "/edit/models").then((json) => {
+            setModels(json);
+        });
+    }, []);
+
+    function handleSelect(event, model) {
+        event.preventDefault();
+        // Ignore categories
+        if (model.slice(0, 2) == "__") {
+            return;
+        }
+        setSelectedModel(model);
+        return false;
+    }
+
+    function handleSubmit(event) {
+        event.preventDefault();
+        const url = ajax.apiUrl + "/edit/object/new";
+        const data = {
+            title: title,
+            noParent: noParent ? 1 : 0,
+            parent: parent,
+            model: selectedModel,
+            state: state,
+        };
+        ajax.ajax({
+            method: "post",
+            url: url,
+            dataType: "json",
+            data: data,
+            error: function (result, status) {
+                setResults("Error! " + result.responseText ?? status);
+            },
+            success: function () {
+                setResults("Success!");
+            },
+        });
+    }
+
+    function handleNoParentChange(event) {
+        setNoParent(event.target.checked);
+        if (allowChangeParentPid && event.target.checked) {
+            setParent("");
+        }
+    }
+
+    function handleParentChange(event) {
+        setParent(event.target.value);
+        if (allowNoParentPid) {
+            setNoParent(event.target.value.length === 0);
+        }
+    }
+
+    function handleStateChange(event) {
+        setState(event.target.value);
+    }
+
+    function handleTitleChange(event) {
+        setTitle(event.target.value);
+    }
+
+    let categories = [];
+    for (let category in models) {
+        let children = [];
+        for (let model in models[category]) {
+            let value = models[category][model];
+            children.push(<TreeItem key={model} nodeId={value} label={model} />);
+        }
+        categories.push(
+            <TreeItem key={category} nodeId={`__${category}`} label={category}>
+                {children}
+            </TreeItem>
+        );
+    }
+    let stateControls = [];
+    states.forEach((currentState) => {
```

Comment:
can do `const stateControls = stats.reduce((currentState) => <label>, []);`

Reply:
On closer examination, I think a map is more appropriate here than a reduce... but in any case, thanks for the suggestion to make this more functional!